adapters/Vungle/VungleAdapter/GADMAdapterVungleRouter.m

```diff
@@ -55,7 +58,7 @@ - (id)init {
     _initializingDelegates = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsStrongMemory
                                                    valueOptions:NSPointerFunctionsWeakMemory];
     _bannerDelegates = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsStrongMemory
-                                             valueOptions:NSPointerFunctionsWeakMemory];
+                                             valueOptions:NSPointerFunctionsStrongMemory];
```

Comment:
This should not be strong. Again, when this class is passed an object and keeps a strong reference to it, it doesn't matter if the caller passed in a __weak poiniter.

Reply:
@ericleich Hi ericleich, we used the weak reference before, but in the testing of multiple Banners with same placementID, `NSPointerFunctionsWeakMemory` always leads the _bannerDelegates becoming empty, because the value object of _bannerDelegates is released early. And our 6.5.2.0 Adapter updates the logic to track Banner delegate lifecycle, after received the callback method `vungleDidCloseAdWithViewInfo:placementID:`, then the adapter will remove the old Banner request delegate from _bannerDelegates, so use strong reference here is fine. And I will remove that redundant __weak assignment operation. Thanks a lot for the instructions. ^_^

Reply:
This is kind of related to the comment I'll send below. I think moving away from a map would will resolve this.

The reason it's getting cleared is because we moved to the value being an ordered set of delegates, not just a single delegate. Back when it was a single delegate, the Google SDK would hold onto the adapter instance if it was still valid, and it would allow the adapter (aka the delegate) to nil out if Google no longer considered this instance valid (such as a request timeout). It's good practice to preserve this behavior via weak references -- ideally Google SDK keeps the only strong reference to the adapter, so that if it gets deallocated (e.g. timeout) and any code path for some reason doesn't call the router to remove the banner delegate, it gets automatically cleared.

Reply:
@ericleich Hi Eric, thanks a lot for your guidance. I have a concern, if we move away from a map and use NSPointerArray with an option for NSPointerFunctionsWeakMemory to manage Banner delegates. During my debugging, I found that AdMob SDK always removes old Banner delegate and set it to nil before Vungle Adapter receives the callback method
`- (void)vungleWillCloseAdWithViewInfo:placementID:` during refreshing process, this could lead the old Banner delegate cannot receive `- (void)willCloseAd:didDownload:` delegate method. This is why we update the logic to track Banner delegate lifecycle, after Vungle Adapter receives the callback method `- (void)vungleDidCloseAdWithViewInfo:placementID:` , the adapter will remove the old Banner delegate from _bannerDelegates. Thanks a lot. :)

Reply:
@YueVungle If Google SDK is dropping the instance, it means we also dropped the ad that the adapter is responding to. So the callback likely isn't doing anything anyways (or if it's making it back, it's only because the adapter didn't let things get cleaned up). I took a look at what `(void)willCloseAd:didDownload:` does and it only seemed to be calling a callback, nothing else that would break functionality.

That said, this discussion may be a moot point if we stick with NSMapTable.

Reply:
@ericleich Hi Eric, thanks a lot for the instructions, I see and I think we should remove the logic to track Banner delegate lifecycle, let Google SDK manage the Banner delegate lifecycle. 😊

Reply:
@ericleich Hi Eric, we changed the `_bannerDelegates` back to use `NSMapTable<GADMAdapterVungleBannerRequest *, id<GADMAdapterVungleDelegate>>` and used `NSPointerFunctionsWeakMemory`, thanks so much for your instruction. 😊